checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt

```diff
@@ -0,0 +1,318 @@
+
+        verifyZeroInteractions(cardValidator)
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with invalid state twice - when cvv is partially valid and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsPartiallyValid()
+
+        createNewAccessCheckoutCVV().onEndUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, false)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state twice - when cvv is complete and valid and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsCompletelyValid()
+
+        createNewAccessCheckoutCVV().onEndUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, true)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with invalid state twice - when cvv is invalid and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsInvalid()
+
+        createNewAccessCheckoutCVV().onEndUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, false)
+    }
+
+    @Test
+    fun `should not raise any events when cardValidator is not set and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsInvalid()
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = null
+
+        accessCheckoutCVV.onEndUpdateCVV(cvvValue)
+
+        verifyZeroInteractions(cardValidator)
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should not raise any update events if there is no cardListener set and cvv is on focus`() {
+        givenCvvIsOnFocus()
+        givenCvvIsPartiallyValid()
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = null
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should not raise any update events if there is no cardListener set and cvv is off focus`() {
+        givenCvvIsOffFocus()
+        givenCvvIsPartiallyValid()
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = null
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should not raise any onUpdateLengthFilter events when cvvLengthFilter is null`() {
+        givenCvvIsOnFocus()
+        givenCvvIsPartiallyValid()
+
+        given(factory.getCVVLengthFilter(cardValidator, null)).willReturn(null)
+
+        createNewAccessCheckoutCVV()
+
+        verifyZeroInteractions(cvvLengthFilter)
+        verify(cardListener, times(0)).onUpdateLengthFilter(cvvView, cvvLengthFilter)
+    }
+
+    private fun createNewAccessCheckoutCVV(): AccessCheckoutCVV {
```

Comment:
[Can merge] - Can be moved after the last test below

Reply:
Done